Translate Refs and the DOM (#52)
reactjs/ja.reactjs.org: content/docs/refs-and-the-dom.md

}
```

コンポーネントがマウントされると React は `current` プロパティに DOM 要素を割り当て、マウントが解除されると `null` を割り当てます。`ref` の更新は `componentDidMount` または `componentDidUpdate` ライフサイクルメソッドの前に行われます。

#### クラスコンポーネントへの Ref の追加

> マウント直後にクリックされることをシミュレーションするために上記の CustomTextInput をラップしたい場合は、ref を使用してカスタムインプットにアクセスし、その focusTextInput メソッドを手動で呼び出せます。

Reviewer: ```suggestion
マウント直後にクリックされることをシミュレーションするために上記の CustomTextInput をラップしたい場合は、ref を使用してカスタムインプットにアクセスし、その `focusTextInput` メソッドを手動で呼び出せます。
```

}
```

これは `CustomTextInput` がクラスとして宣言されている場合にのみ機能することに注意してください。

```js{1}
class CustomTextInput extends React.Component {
  // ...
}
```

> #### Refs と Function Components

Reviewer: ```suggestion
#### Refs と 関数コンポーネント
```

> Function components にはインスタンスがないため、**Function components に `ref` 属性を使用することはできません。**

Reviewer: ```suggestion
関数コンポーネントにはインスタンスがないため、**関数コンポーネントに `ref` 属性を使用することはできません。**
```

    this.textInput = React.createRef();
  }
  render() {

>     // これは動き*ません*！

Reviewer 2: ```suggestion
    // これは動き「ません」！
```

コード内なので

    return (
      <MyFunctionComponent ref={this.textInput} />
    );
  }
}
```

> Ref が必要な場合は、ライフサイクルメソッドやステートが必要なときと同じように、コンポーネントをクラスに変換しなければなりません。

Reviewer: ```suggestion
Ref が必要な場合は、ライフサイクルメソッドや state が必要なときと同じように、コンポーネントをクラスに変換しなければなりません。
```

> ただし、DOM 要素またはクラスコンポーネントを参照している限り、**関数コンポーネント内で ref 属性を使用できます。**

Reviewer 2: ```suggestion
ただし、DOM 要素またはクラスコンポーネントを参照している限り、**関数コンポーネント内で ref 属性を使用することはできます。**
```

> ただし、DOM 要素またはクラスコンポーネントを参照している限り、**function component 内で ref 属性を使用できます。**

Reviewer: ```suggestion
ただし、DOM 要素またはクラスコンポーネントを参照している限り、**関数コンポーネント内で ref 属性を使用できます。**
```

}
```

> ### DOM の Refs を親コンポーネントにさらす

Reviewer: ```suggestion
### DOM の Refs を親コンポーネントに公開する
```

Reviewer: 「公開する」も逐語訳っぽくて、微妙なので意見求む 🙏

Author: 確かにここは自分も「公開する」にしようか悩みました🤔
他の良い表現が見つかりませんが、「公開する」のほうが適切な気がします👍

> まれに、親コンポーネントから子の DOM ノードにアクセスしたい場合があります。 これは、コンポーネントのカプセル化を壊すため、一般的にはおすすめできませんが、フォーカスを発火したり、子の DOM ノードのサイズや位置を計測したりするのに役立つことがあります。

Reviewer: ```suggestion
まれに、親コンポーネントから子の DOM ノードにアクセスしたい場合があります。これは、コンポーネントのカプセル化を壊すため、一般的にはおすすめできませんが、フォーカスを発火させたり、子の DOM ノードのサイズや位置を計測するのに役立つことがあります。
```

> [子コンポーネントに ref を追加すること](#adding-a-ref-to-a-class-component)はできますが、DOM ノードではなくコンポーネントインスタンスしか取得できないため、これは理想的な解決策ではありません。 また、これは function components では機能しません。

Reviewer: ```suggestion
[子コンポーネントに ref を追加すること](#adding-a-ref-to-a-class-component)はできますが、DOM ノードではなくコンポーネントインスタンスしか取得できないため、これは理想的な解決策ではありません。また、これは関数コンポーネントでは機能しません。
```

> React 16.3 以降を使用している場合、これらの場合には [ref forwarding](/docs/forwarding-refs.html) を使用することをおすすめします。Ref forwarding では、コンポーネントは子コンポーネントの参照を自分のものとして公開することを選択できます。 [Ref forwarding のドキュメント](/docs/forwarding-refs.html#forwarding-refs-to-dom-components)に、子の DOM ノードを親コンポーネントに公開する方法の詳細な例があります。

Reviewer: ```suggestion
React 16.3 以降を使用している場合、これらの場合には [ref forwarding](/docs/forwarding-refs.html) を使用することをおすすめします。**Ref forwarding では、コンポーネントは任意の子コンポーネントの ref を自分のものとして公開することを選択できます。** [Ref forwarding のドキュメント](/docs/forwarding-refs.html#forwarding-refs-to-dom-components)に、子の DOM ノードを親コンポーネントに公開する方法の詳細な例があります。
```

> React 16.2 以下を使用している場合、または ref forwarding で提供される以上の柔軟性が必要な場合は、[この代替アプローチ](https://gist.github.com/gaearon/1a018a023347fe1c2476073330cc5509)を使用して ref を別の名前の prop として明示的に渡すことができます。

Reviewer: ```suggestion
React 16.2 以下を使用している場合、または ref forwarding で提供される以上の柔軟性が必要な場合は、[この代替手法](https://gist.github.com/gaearon/1a018a023347fe1c2476073330cc5509)を使用して ref を異なる名前の props として明示的に渡すことができます。
```

> 可能であれば DOM ノードをさらさないことをおすすめしますが、これは便利な回避策になることもあります。この方法では、子コンポーネントにコードを追加する必要があります。子コンポーネントの実装を完全に制御できない場合、最後の選択肢は [`findDOMNode()`](/docs/react-dom.html#finddomnode) を使用することですが、[`StrictMode`](/docs/strict-mode.html#warning-about-deprecated-finddomnode-usage) では推奨されません。

Reviewer: ```suggestion
可能であれば DOM ノードを公開しないことをおすすめしますが、これは便利な避難ハッチになることもあります。留意すべき点として、この方法では子コンポーネントにコードを追加する必要があります。子コンポーネントの実装にまったく手を加えられない場合、最後の選択肢は [`findDOMNode()`](/docs/react-dom.html#finddomnode) を使用することですが、
おすすめできない上に、[`StrictMode`](/docs/strict-mode.html#warning-about-deprecated-finddomnode-usage) では
廃止予定です。
```

### コールバック Refs

React はまた「コールバック Refs」と呼ばれる refs を設定するためのさらなる方法をサポートします。

> `createRef()` によって作成された `ref` 属性を渡す代わりに、関数を渡します。 この関数は、引数として React コンポーネントのインスタンスまたは HTML DOM 要素を受け取ります。これらは他の場所に保持してアクセスできます。

Reviewer: ```suggestion
`createRef()` によって作成された `ref` 属性を渡す代わりに、関数を渡します。 この関数は、引数として React コンポーネントのインスタンスまたは HTML DOM 要素を受け取ります。これらは他の場所に保持したり、他の場所からアクセスできます。
```

}
```

> ### DOM の Ref を親コンポーネントにさらす

Reviewer 2: ```suggestion
### DOM の Ref を親コンポーネントに公開する
```

> まれに、親コンポーネントから子の DOM ノードにアクセスしたい場合があります。これは、コンポーネントのカプセル化を壊すため、一般的にはおすすめできませんが、フォーカスを発火させたり、子の DOM ノードのサイズや位置を計測するのに役立つことがあります。

Reviewer 2: ```suggestion
まれに、親コンポーネントから子コンポーネントの DOM ノードにアクセスしたい場合があります。これは、コンポーネントのカプセル化を壊すため、一般的にはおすすめできませんが、フォーカスを発火させたり、子の DOM ノードのサイズや位置を計測するのに役立つことがあります。
```

「子のDOMノード」だと、単に親が直接renderしているDOMノードのことだと混同しそうなので、コンポーネントが2種類登場することを明記しました

[子コンポーネントに ref を追加すること](#adding-a-ref-to-a-class-component)はできますが、DOM ノードではなくコンポーネントインスタンスしか取得できないため、これは理想的な解決策ではありません。また、これは関数コンポーネントでは機能しません。

> React 16.3 以降を使用している場合、これらの場合には [ref forwarding](/docs/forwarding-refs.html) を使用することをおすすめします。**Ref forwarding では、コンポーネントは任意の子コンポーネントの ref を自分のものとして公開することを選択できます。** [Ref forwarding のドキュメント](/docs/forwarding-refs.html#forwarding-refs-to-dom-components)に、子の DOM ノードを親コンポーネントに公開する方法の詳細な例があります。

Reviewer 2: ```suggestion
React 16.3 以降を使用している場合、これらの場合には [ref forwarding](/docs/forwarding-refs.html) を使用することをおすすめします。**Ref forwarding では、コンポーネントは任意の子コンポーネントの ref を自分のものとして公開することを選択できます。**[Ref forwarding のドキュメント](/docs/forwarding-refs.html#forwarding-refs-to-dom-components)に、子の DOM ノードを親コンポーネントに公開する方法の詳細な例があります。
```

React 16.2 以下を使用している場合、または ref forwarding で提供される以上の柔軟性が必要な場合は、[この代替手法](https://gist.github.com/gaearon/1a018a023347fe1c2476073330cc5509)を使用して ref を異なる名前の props として明示的に渡すことができます。

可能であれば DOM ノードを公開しないことをおすすめしますが、これは便利な避難ハッチになることもあります。留意すべき点として、この方法では子コンポーネントにコードを追加する必要があります。子コンポーネントの実装にまったく手を加えられない場合、最後の選択肢は [`findDOMNode()`](/docs/react-dom.html#finddomnode) を使用することですが、

> おすすめできない上に、[`StrictMode`](/docs/strict-mode.html#warning-about-deprecated-finddomnode-usage) では

Reviewer 2: これらの行を上とつなげてください（改行は半角スペースになるのでレンダリングにも影響を与えます）

廃止予定です。

### コールバック Ref

> React はまた「コールバック Ref」と呼ばれる ref を設定するためのさらなる方法をサポートします。

Reviewer 2: ```suggestion
React はまた「コールバック Ref」と呼ばれる ref を設定するための別の方法をサポートします。
```

    };

    this.focusTextInput = () => {

>       // 生の DOM API を使用してテキストの入力にフォーカスします。

Reviewer: ```suggestion
      // 生の DOM API を使用して明示的にテキストの入力にフォーカスします。
```

}
```

> コンポーネントがマウントされると React は DOM 要素で `ref` コールバックを呼び出し、マウントが解除されると `null` を呼び出します。 Refs は `componentDidMount` または `componentDidUpdate` が発火される前に最新のものであることが保証されています。

Reviewer: ```suggestion
コンポーネントがマウントされると React は DOM 要素とともに `ref` コールバックを呼び出し、マウントが解除されると `null` とともにコールバックを呼び出します。Refs は `componentDidMount` または `componentDidUpdate` が発火する前に最新のものであることが保証されています。
```

}
```

> 上記の例では、`Parent` は ref コールバックを `inputRef` prop として `CustomTextInput` に渡し、`CustomTextInput` は同じ関数を特別な `ref` 属性として `<input>` に渡します。 その結果、`Parent` の `this.inputElement` は、`CustomTextInput` の `<input>` 要素に対応する DOM ノードに設定されます。

Reviewer: ```suggestion
上記の例では、`Parent` は ref コールバックを `inputRef` プロパティ として `CustomTextInput` に渡し、`CustomTextInput` は同じ関数を特別な `ref` 属性として `<input>` に渡します。その結果、`Parent` の `this.inputElement` は、`CustomTextInput` の `<input>` 要素に対応する DOM ノードに設定されます。
```

> ### レガシー API：String Refs

Reviewer: ```suggestion
### レガシー API：文字列 Refs
```

Reviewer: String 訳さなくてもいいかも 🤔

Author: 個人的には訳さなくても良いかと思います！

以前に React を使用したことがある場合は、`ref` 属性が `"textInput"` のような文字列で、DOM ノードが `this.refs.textInput` としてアクセスされる古い API に慣れているかもしれません。String ref には[いくつかの問題](https://github.com/facebook/react/pull/8333#issuecomment-271648615)があり、レガシーと見なされ、**将来のリリースのいずれかで削除される可能性が高い**ため、使用することをおすすめしません。

> 補足
>
> Ref にアクセスするために `this.refs.textInput` を現在使用している場合は、代わりに[コールバックパターン](#callback-refs)もしくは [`createRef` API](#creating-refs) を使用することをおすすめします。

> ### Callback ref の注意事項

Reviewer 3: ```suggestion
### コールバック Ref の注意事項
```

> `ref` コールバックがインライン関数として定義されている場合、更新中に 2 回呼び出されます。最初は `null`、次に DOM 要素で呼び出されます。これは、各レンダリングで関数の新しいインスタンスが作成されるため、React は古い ref を削除し、新しい ref を設定する必要があるためです。`ref` コールバックをクラスのバウンドメソッドとして定義することでこれを回避できますが、ほとんどの場合は問題にならないはずです。

Reviewer: ```suggestion
`ref` コールバックがインライン関数として定義されている場合、更新中に 2 回呼び出されます。最初は `null`、次に DOM 要素で呼び出されます。これは、それぞれのレンダーで関数の新しいインスタンスが作成されるため、React は古い ref を削除し、新しい ref を設定する必要があるためです。`ref` コールバックをクラスのバウンドメソッドとして定義することでこれを回避できますが、ほとんどの場合は問題にならないはずです。
```

}
```

上記の例では、`Parent` は ref コールバックを `inputRef` プロパティとして `CustomTextInput` に渡し、`CustomTextInput` は同じ関数を特別な `ref` 属性として `<input>` に渡します。その結果、`Parent` の `this.inputElement` は、`CustomTextInput` の `<input>` 要素に対応する DOM ノードに設定されます。

### レガシー API：String Ref

以前に React を使用したことがある場合は、`ref` 属性が `"textInput"` のような文字列で、DOM ノードが `this.refs.textInput` としてアクセスされる古い API に慣れているかもしれません。String ref には[いくつかの問題](https://github.com/facebook/react/pull/8333#issuecomment-271648615)があり、レガシーと見なされ、**将来のリリースのいずれかで削除される可能性が高い**ため、使用することをおすすめしません。

> 補足
>
> Ref にアクセスするために `this.refs.textInput` を現在使用している場合は、代わりに[コールバックパターン](#callback-refs)もしくは [`createRef` API](#creating-refs) を使用することをおすすめします。

### Callback ref の注意事項

> `ref` コールバックがインライン関数として定義されている場合、更新中に 2 回呼び出されます。最初は `null`、次に DOM 要素で呼び出されます。これは、それぞれのレンダーで関数の新しいインスタンスが作成されるため、React は古い ref を削除し、新しい ref を設定する必要があるためです。`ref` コールバックをクラスのバウンドメソッドとして定義することでこれを回避できますが、ほとんどの場合は問題にならないはずです。

Reviewer 2: ```suggestion
`ref` コールバックがインライン関数として定義されている場合、更新中に 2 回呼び出されます。最初は `null`、次に DOM 要素で呼び出されます。これは、それぞれのレンダーで関数の新しいインスタンスが作成されるため、React は古い ref を削除し、新しい ref を設定する必要があるためです。`ref` コールバックをクラス内のバインドされたメソッドとして定義することでこれを回避できますが、ほとんどの場合は問題にならないはずです。
```
